nixos/modules/services/hardware/scanservjs.nix

```
@@ -0,0 +1,155 @@
+{
+  config,
+  lib,
+  pkgs,
+  ...
+}:
+
+let
+  cfg = config.services.scanservjs;
+  settings = {
+    scanimage = "${config.hardware.sane.backends-package}/bin/scanimage";
+    convert = "${pkgs.imagemagick}/bin/convert";
+    tesseract = lib.getExe pkgs.tesseract;
+    # it defaults to config/devices.json, but "config" dir doesn't exist by default and scanservjs doesn't create it
+    devicesPath = "devices.json";
+  } // cfg.settings;
+  settingsFormat = pkgs.formats.json { };
+
+  leafs =
+    attrs:
+    builtins.concatLists (
+      lib.mapAttrsToList (k: v: if builtins.isAttrs v then leafs v else [ v ]) attrs
+    );
+
+  package = pkgs.scanservjs;
+
+  configFile = pkgs.writeText "config.local.js" ''
+    /* eslint-disable no-unused-vars */
+    module.exports = {
+      afterConfig(config) {
+        ${
+          builtins.concatStringsSep "" (
+            leafs (
+              lib.mapAttrsRecursive (path: val: ''
+                ${builtins.concatStringsSep "." path} = ${builtins.toJSON val};
+              '') { config = settings; }
+            )
+          )
+        }
+        ${cfg.extraConfig}
+      },
+
+      afterDevices(devices) {
+        ${cfg.extraDevicesConfig}
+      },
+
+      async afterScan(fileInfo) {
+        ${cfg.runAfterScan}
+      },
+
+      actions: [
+        ${builtins.concatStringsSep ",\n" cfg.extraActions}
+      ],
+    };
+  '';
+
+in
+{
+  options.services.scanservjs = {
+    enable = lib.mkEnableOption "scanservjs";
+    stateDir = lib.mkOption {
+      type = lib.types.str;
+      default = "/var/lib/scanservjs";
+      description = ''
+        State directory for scanservjs
+      '';
+    };
+    settings = lib.mkOption {
+      default = { };
+      description = ''
+        Config to set in config.local.js's `afterConfig`
+      '';
+      type = lib.types.submodule {
+        freeformType = settingsFormat.type;
+        options.host = lib.mkOption {
+          type = lib.types.str;
+          description = "The IP to listen on";
+          default = "127.0.0.1";
+        };
+        options.port = lib.mkOption {
+          type = lib.types.port;
+          description = "The port to listen on";
+          default = 8080;
+        };
+      };
+    };
+    extraConfig = lib.mkOption {
+      default = "";
+      type = lib.types.lines;
+      description = ''
+        Extra code to add to config.local.js's `afterConfig`
```

```suggestion
        Extra code to add to config.local.js's `afterConfig`.
```
I am not sure how useful this is given the settings option already exists.

---

this is potentially useful for running arbitrary js